src/libmeasurement_kit/ooni/whatsapp.cpp

```diff
@@ -0,0 +1,211 @@
+                logger->info("tcp failure to %s:%d", ip.c_str(), port);
+                close_txp = false;
+            } else {
+                logger->info("tcp success to %s:%d", ip.c_str(), port);
+            }
+            *ips_tested += 1;
+            if (ips_count == *ips_tested) {
+                if (close_txp == true) {
+                    txp->close([=] { cb(NoError()); });
+                } else {
+                    cb(NoError());
+                }
+            } else {
+                if (close_txp == true) {
+                    // XXX optimistic closure
+                    txp->close([=] {});
+                }
+            }
+        };
+    };
+
+    for (auto const& ip : ips) {
+        // XXX hardcoded
+        std::vector<int> ports {443, 5222};
+        for (auto const& port : ports) {
+            Settings tcp_options;
+            tcp_options["host"] = ip;
+            tcp_options["port"] = port;
+            tcp_options["net/timeout"] = 10.0; //XXX hardcoded
+            templates::tcp_connect(tcp_options, tcp_cb(ip, port),
+                                   reactor, logger);
+        }
+    }
+    return;
+}
+
+static void dns_many(std::vector<std::string> hostnames,
+                     Var<Entry> entry,
+                     Callback<Error, std::vector<std::string>> cb,
+                     Var<Reactor> reactor,
+                     Var<Logger> logger) {
+    int names_count = hostnames.size();
+    logger->info("whatsapp: %d hostnames", names_count);
+    Var<std::vector<std::string>> ips(new std::vector<std::string>);
+    Var<int> names_tested(new int(0));
+    if (names_count == *names_tested) {
+        cb(NoError(), *ips);
+    }
+
+
+    auto dns_cb = [=](std::string hostname) {
+        return [=](Error err, Var<dns::Message> message) {
+            if (!!err) {
+                logger->info("whatsapp: dns error for %s", hostname.c_str());
+            } else {
+                for (auto answer : message->answers) {
+                    // XXX i don't get this answer format
+                    // copying from web_connectivity
+                    if (answer.ipv4 != "") {
+                        logger->info("(1) %s ipv4: %s",
+                            hostname.c_str(), answer.ipv4.c_str());
+                        ips->push_back(answer.ipv4);
+                    } else if (answer.hostname != "") {
+                        logger->info("(2) %s ipv4: %s",
+                            hostname.c_str(), answer.ipv4.c_str());
+                        ips->push_back(answer.ipv4);
+                    }
+                }
+            }
+            *names_tested += 1;
+            if (names_count == *names_tested) {
+                logger->info("dns_many() found %d ips", (int)ips->size());
+                cb(NoError(), *ips);
+            }
+        };
+    };
+
+    for (auto const& hostname : hostnames) {
+        Settings options; // XXX ought this to forward from above?
```

Comment:
Yep, I guess this should be the options from above. AFAICT, you don't need to worry about clobbering options because in MK options are always passed by copy to functions precisely for this reason.

Reply:
You can do `Settings local_options = options;` to make a copy in a in-loop variable.